File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        m,n = len(s),len(t)
        if m < n:
            return ""
        need = [0] * 256
        for ch in t:
            need[ord(ch)] += 1

        missing = n
        l = 0
        INF = int(1e5)+1 # bigger than any possible len
        bestLen = INF
        bestStart = 0
        for r in range(m):
            cr = s[r]

            if need[ord(cr)] > 0:
                missing -= 1
            
            need[ord(cr)] -= 1

            while missing == 0:
                winLen = r-l+1
                if winLen < bestLen:
                    bestLen = winLen
                    bestStart = l
                cl = s[l]
                need[ord(cl)] += 1
                if need[ord(cl)] > 0:
                    missing += 1
                l += 1

        return "" if bestLen == INF else s[bestStart:bestStart+bestLen]

File: 0076-minimum-window-substring/0076-minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        int m = (int)s.size();
        int n = (int)t.size();
        if (m < n)
            return "";
        int need[58];
        for (char ch : t)
            need[ch-'A']++;
        int missing = n;
        int bestStart = 0;
        int bestLen = INT_MAX;
        int l = 0;
        for (int r=0; r<m; r++){
            char cr = s[r];
            if (need[cr-'A'] > 0)
                missing--;
            need[cr-'A']--;
            while (missing == 0){
                int len = r-l+1;
                if (len < bestLen){
                    bestLen = len;
                    bestStart = l;
                }
                int cl = s[l];
                need[cl-'A']++;
                if (need[cl-'A'] > 0)
                    missing++;
                l++;
            }
        }
        return (bestLen == INT_MAX) ? "" : s.substr(bestStart,bestLen);
    }
};